docs/architecture/adr-versions.md

```diff
@@ -0,0 +1,219 @@
+cluttered with conditionals to effect different rules at different heights.
+
+More details on how this manager process works later below.
+
+### ChainDescription
+
+ChainDescription is a complete immutable description of a blockchain. It takes the following form:
+
+```
+ChainDescription = <NetworkName>/<BlockVersion>/<AppVersion>/<StateHash>/<ValHash>/<ConsensusParamsHash>
+```
+
+Here, StateHash is the merkle root of the initial state, ValHash is the merkle root of the initial Tendermint validator set,
+and ConsensusParamsHash is the merkle root of the initial Tendermint consensus parameters.
+
+The `genesis.json` file must contain enough information to compute this value. It need not contain the StateHash or ValHash itself,
+but contain the state from which they can be computed with the given protocol versions.
+
+XXX: should we split NetworkName into NetworkName and AppName?
+
+#### ChainID
+
+Define `ChainID = TMHASH(ChainDescriptor)`. It's the unique ID of a blockchain.
+
+It should be Bech32 encoded when handled by users, eg. with `cosmoschain` prefix.
+
+#### Forks and Uprades
+
+When a chain forks or upgrades but continues the same history, it takes a new ChainDescription as follows:
+
+```
+ChainDescription = <ChainID>/x/<Height>/<ForkDescription>
+```
+
+Where
+    - ChainID is the ChainID from the previous ChainDescription (ie. its hash)
+    - `x` denotes that a change occured
+    - `Height` is the height the change occured
+    - ForkDescription has the same form as ChainDescription but for the fork
+        - this allows forks to specify new versions for tendermint or the app, as well as arbitrary changes to the state or validator set
+
+### ABCI Changes
+
+On RequestInfo, Tendermint should send: TMSoftwareVersion, BlockVersion, AppVersion
+On ResponseInfo, App should send: AppSoftwareVersion, AppVersion
+
+For the App to signal an upgrade, it should use the following tags (KVPairs) in ResponseEndBlock:
+
+- `("upgrade_chain_description", <new ChainDescription>)`
+- `("upgrade_chain_height", <height to upgrade>)`
+
+This provides a simple manner for applications to signal to external processes that they want to upgrade, and what height to do it at.
+
+By using the ChainDescription, the app signals a summary of all changes expected to be made and when and exactly how the ChainDescription will change.
+
+### Tendermint Changes
+
+Compatibility requirements at the peer layer have already been described.
+
+The only other change necessary is an `/unsafe_stop?height=_` endpoint to tell Tendermint to shutdown at a given height.
+This is to enable it to be stopped at the height signalled by `upgrade_chain_height` and to then be restarted using the new ChainDescription (ie. new app).
+Note if the ValidatorHash or ConsensusParamsHash were force-changed, Tendermint will require the new information.
+
+
+### Process Manager
+
+A simple process managers subscribes to the `upgrade_chain` events and calls `/unsafe_stop` for the given height.
```

Comment:
Just to be clear, this is to avoid hard-coded conditionals on upgrades via app versions correct? Will this just be a service that runs along side a Tendermint node?